File: model/src/main/java/no/nav/pensjon/vtp/testmodell/repo/impl/JournalRepositoryImpl.java
package no.nav.pensjon.vtp.testmodell.repo.impl;

import static java.util.Optional.ofNullable;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

import org.springframework.stereotype.Repository;

import no.nav.pensjon.vtp.testmodell.dokument.modell.DokumentModell;
import no.nav.pensjon.vtp.testmodell.dokument.modell.JournalpostModell;
import no.nav.pensjon.vtp.testmodell.repo.JournalRepository;

@Repository
public class JournalRepositoryImpl implements JournalRepository {

    private final HashMap<String, JournalpostModell> journalposter;
    private final HashMap<String, DokumentModell> dokumenter;

    private final AtomicInteger journalpostId;
    private final AtomicInteger dokumentId;

    public JournalRepositoryImpl() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("Mdkm");
        journalposter = new HashMap<>();
        dokumenter = new HashMap<>();
        journalpostId = new AtomicInteger(Integer.parseInt(LocalDateTime.now().format(formatter)) * 100);
        dokumentId = new AtomicInteger(Integer.parseInt(LocalDateTime.now().format(formatter)) * 100);
    }

    @Override
    public Optional<DokumentModell> finnDokumentMedDokumentId(String dokumentId) {
        return ofNullable(dokumenter.get(dokumentId));
    }

    @Override
    public List<JournalpostModell> finnJournalposterMedFnr(String fnr){
        return journalposter.values().stream()
                .filter(e -> (e.getAvsenderFnr() != null && e.getAvsenderFnr().equalsIgnoreCase(fnr)))
                .collect(Collectors.toList());
    }

    @Override
    public List<JournalpostModell> finnJournalposterMedSakId(String sakId){
        return journalposter.values().stream()
                .filter(e -> (e.getSakId() != null && e.getSakId().equalsIgnoreCase(sakId)))
                .collect(Collectors.toList());
    }

    @Override
    public Optional<JournalpostModell> finnJournalpostMedJournalpostId(String journalpostId){
        return ofNullable(journalposter.get(journalpostId));
    }

    @Override
    public String leggTilJournalpost(JournalpostModell journalpostModell){
        String journalpostId;
        if(journalpostModell.getJournalpostId() != null && !journalpostModell.getJournalpostId().isEmpty()){
            journalpostId = journalpostModell.getJournalpostId();
        } else {
            journalpostId = genererJournalpostId();
            journalpostModell.setJournalpostId(journalpostId);
        }

        for(DokumentModell dokumentModell : journalpostModell.getDokumentModellList()){
            String dokumentId;
            if(dokumentModell.getDokumentId() != null && !journalpostModell.getJournalpostId().isEmpty()){
                dokumentId = dokumentModell.getDokumentId();
            } else {
                dokumentId = genererDokumentId();
                dokumentModell.setDokumentId(dokumentId);
            }
            if(dokumenter.containsKey(dokumentId)){
                throw new IllegalStateException("Forsøker å opprette dokument allerede eksisterende dokumentId");
            } else {
                dokumenter.put(dokumentId, dokumentModell);
            }
        }

        if(journalposter.containsKey(journalpostId)){
            throw new IllegalStateException("Forsøker å opprette journalpost allerede eksisterende journalpostId");
        } else {
            journalposter.put(journalpostId, journalpostModell);
            return journalpostId;
        }
    }


    public String genererJournalpostId(){
        return Integer.toString(journalpostId.incrementAndGet());
    }

    public String genererDokumentId(){
        return Integer.toString(dokumentId.incrementAndGet());
    }




}

File: vtp-pensjon-application/src/main/kotlin/no/nav/pensjon/vtp/testmodell/repo/impl/JournalRepositoryImpl.kt
package no.nav.pensjon.vtp.testmodell.repo.impl

import no.nav.pensjon.vtp.testmodell.dokument.modell.DokumentModell
import no.nav.pensjon.vtp.testmodell.dokument.modell.JournalpostModell
import no.nav.pensjon.vtp.testmodell.repo.JournalRepository
import org.springframework.stereotype.Repository
import java.time.LocalDateTime.now
import java.time.format.DateTimeFormatter.ofPattern
import java.util.concurrent.atomic.AtomicInteger

@Repository
class JournalRepositoryImpl : JournalRepository {
    private val journalposter = HashMap<String, JournalpostModell>()
    private val dokumenter = HashMap<String, DokumentModell>()

    private val journalpostId = AtomicInteger(now().format(ofPattern("Mdkm")).toInt() * 100)
    private val dokumentId = AtomicInteger(now().format(ofPattern("Mdkm")).toInt() * 100)

    override fun finnDokumentMedDokumentId(dokumentId: String): DokumentModell? {
        return dokumenter[dokumentId]
    }

    override fun finnJournalposterMedFnr(fnr: String): List<JournalpostModell> {
        return journalposter.values
            .filter { e: JournalpostModell -> e.avsenderFnr != null && e.avsenderFnr.equals(fnr, ignoreCase = true) }
    }

    override fun finnJournalposterMedSakId(sakId: String): List<JournalpostModell> {
        return journalposter.values
            .filter { it.sakId != null && it.sakId.equals(sakId, ignoreCase = true) }
    }

    override fun finnJournalpostMedJournalpostId(journalpostId: String): JournalpostModell? {
        return journalposter[journalpostId]
    }

    override fun finnJournalpostMedDokumentId(dokumentId: String): JournalpostModell? =
        journalposter.values
            .firstOrNull { it.dokumentModellList.any { dok -> dok.dokumentId == dokumentId } }

    override fun save(journalpostModell: JournalpostModell): JournalpostModell {
        val journalpostId: String = journalpostModell.journalpostId?.let {
            if (it.isNotEmpty()) {
                it
            } else {
                null
            }
        }
            ?: genererJournalpostId()

        journalpostModell.journalpostId = journalpostId

        for (dokumentModell in journalpostModell.dokumentModellList) {
            val dokumentId =
                dokumentModell.dokumentId?.let {
                    if (it.isNotEmpty()) {
                        it
                    } else {
                        null
                    }
                }
                    ?: genererDokumentId()

            dokumentModell.dokumentId = dokumentId

            check(!dokumenter.containsKey(dokumentId)) { "Forsøker å opprette dokument allerede eksisterende dokumentId" }
            dokumenter[dokumentId] = dokumentModell
        }

        return if (journalposter.containsKey(journalpostId)) {
            throw IllegalStateException("Forsøker å opprette journalpost allerede eksisterende journalpostId")
        } else {
            journalposter[journalpostId] = journalpostModell
            journalpostModell
        }
    }

    fun genererJournalpostId(): String {
        return journalpostId.incrementAndGet().toString()
    }

    fun genererDokumentId(): String {
        return dokumentId.incrementAndGet().toString()
    }
}
